js/app.js

```diff
@@ -36,65 +42,107 @@ const App = {
 			`
 			);
 		},
+		getTodo(liOrId) {
+			let $li = liOrId instanceof Element ? liOrId :
+				document.querySelector(`[data-id="${liOrId}"]`);
+			if ($li != null) {
+				return {
+					$li,
+					get $input() {
+						return $li.querySelector('[data-todo="edit"]')
+					},
+					get $label() {
+						return $li.querySelector('[data-todo="label"]')
+					}
+				}
+			}
+		},
+		getEditing() {
+			const $li = document.querySelector('.editing');
+			if ($li != null) {
+				const editingId = $li.dataset.id;
+				const $input = App.$.getTodo($li).$input;
+				const selection = [$input.selectionStart, $input.selectionEnd];
+				return {
+					id: editingId,
+					restore() { return App.$.setEditing(editingId, ...selection); }
+				}
+			}
+		},
+		setEditing(liOrId, selectionStart, selectionEnd) {
+			let $todo = App.$.getTodo(liOrId);
+			if ($todo != null) {
+				$todo.$li.classList.add("editing");
+				$todo.$input.focus();
+				$todo.$input.setSelectionRange(selectionStart, selectionEnd);
+				return $todo;
+			}
+		}
 	},
 	init() {
-		Todos.addEventListener("save", App.render);
-		App.filter = getURLHash();
-		window.addEventListener("hashchange", () => {
-			App.filter = getURLHash();
-			App.render();
-		});
+		function onHashChange() {
+			let {todosId, filter} = getURLHash(document.location.hash);
+			const isNew = !todosId;
+			if (isNew) {
+				todosId = uuid();
+				history.pushState(null, null, `#/${todosId}/${filter}`);
+			}
+			if (App.todos == null || App.todos.id !== todosId) {
+				App.todos?.close();
+				App.todos = new TodoStore(todosId, isNew);
+				App.$.updateFilterHashes();
+				App.todos.addEventListener("save", App.render);
+				App.todos.addEventListener("error", App.error);
+			}
+			App.filter = filter;
+		}
```

Comment:
**issue:** The app currently doesn't render when the filter changes, so the
filter buttons appear to do nothing. How about this?

```suggestion
			if (App.filter !== filter) {
				App.filter = filter;
				App.render();
			}
		}
```